### ISSUE TITLE
Mysql2::Error: Lost connection to MySQL server at 'reading initial communication packet', system error: 102

### ISSUE THREAD

**Issue by joshuapinter** [brianmario/mysql2#1000]
Hi there, 

I tried solving this myself but after countless weeks I just couldn't find a stable solution so I thought I would ask here to get more insight.

Generally, the `mysql2` gem works brilliantly with our Rails 4.0 app. We recently started using the [`parallel_tests` gem](https://github.com/grosser/parallel_tests) to make our tests multi-threaded and speed up our test suite.

However, when running this locally on my MacBook Pro with 4-cores (8 threads), I will sometimes get the following error: 

```
Mysql2::Error: Lost connection to MySQL server at 'reading initial communication packet', system error: 102
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/mysql2-0.3.21/lib/mysql2/client.rb:70:in `connect'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/mysql2-0.3.21/lib/mysql2/client.rb:70:in `initialize'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_adapters/mysql2_adapter.rb:18:in `new'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_adapters/mysql2_adapter.rb:18:in `mysql2_connection'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_adapters/abstract/connection_pool.rb:440:in `new_connection'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_adapters/abstract/connection_pool.rb:450:in `checkout_new_connection'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_adapters/abstract/connection_pool.rb:421:in `acquire_connection'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_adapters/abstract/connection_pool.rb:356:in `block in checkout'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_adapters/abstract/connection_pool.rb:355:in `checkout'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_adapters/abstract/connection_pool.rb:265:in `block in connection'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_adapters/abstract/connection_pool.rb:264:in `connection'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_adapters/abstract/connection_pool.rb:546:in `retrieve_connection'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_handling.rb:79:in `retrieve_connection'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/activerecord-4.0.13/lib/active_record/connection_handling.rb:53:in `connection'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/database_cleaner-1.5.3/lib/database_cleaner/active_record/truncation.rb:234:in `clean'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/database_cleaner-1.5.3/lib/database_cleaner/base.rb:46:in `clean_with'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/database_cleaner-1.5.3/lib/database_cleaner/configuration.rb:91:in `block in clean_with'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/database_cleaner-1.5.3/lib/database_cleaner/configuration.rb:91:in `each'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/database_cleaner-1.5.3/lib/database_cleaner/configuration.rb:91:in `clean_with'
  /Users/me/Development/cntral/spec/spec_helper.rb:30:in `<top (required)>'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/configuration.rb:1455:in `require'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/configuration.rb:1455:in `block in requires='
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/configuration.rb:1455:in `each'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/configuration.rb:1455:in `requires='
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/configuration_options.rb:112:in `block in process_options_into'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/configuration_options.rb:111:in `each'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/configuration_options.rb:111:in `process_options_into'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/configuration_options.rb:21:in `configure'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:99:in `setup'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:86:in `run'
  /Users/me/.rbenv/versions/2.3.6/lib/ruby/gems/2.3.0/gems/rspec-core-3.7.1/lib/rspec/core/runner.rb:71:in `run'
```

Once this occurs, I have to quit the mysql daemon and start it again before it begins to work again.

I assume it has to do with multiple connections to the database but because it doesn't happen all the time and not immediately after a mysqld restart, I'm unsure what it is. 

I've played around with mysql settings in `my.cnf`, including `max_connect_errors` and `max_connections` but nothing seems to help.

Let me know what you think and maybe we can get this solved. Thanks!

### Versions

```ruby
gem "mysql2", "0.3.21"
gem "rails",  "4.0.13"
```

```bash
$ mysql --version
mysql  Ver 14.14 Distrib 5.7.17, for macos10.12 (x86_64) using  EditLine wrapper
```

**Comment by baljeet03** [comment]
I faced the similar problem when I installed mysql 5.7.17 first time. That time I was new to development. Few of the things you can try is look into the error log and general logs itself. 
From what it seems, it might be problem with number of tables getting opened. 
Try setting the `table_open_cache` to twice of max_connections. 
BTW what is your `max_connections`. And also increase the `wait_timeout` and `connect_timeout`

**Comment by joshuapinter** [comment]
Thanks for your response, @baljeet03!

## tldr; 

I had to set my `table_open_cache` to something smaller than the default value of `2000`, like `400`.


## Full Story

First off, my current value for `max_connections`: 

```bash
mysql> show variables like "max_connections";
+-----------------+--------+
| Variable_name   | Value  |
+-----------------+--------+
| max_connections | 100000 |
+-----------------+--------+
1 row in set (0.00 sec)
```

I tried fruitlessly to increase this to `200000` buy it wouldn't increase. Thinking I had done something wrong I did some more digging and found out that `100000` is actually the maximum value. 😳

Now, you've mentioned a few configuration variables that I haven't seen before. Here were the current values before making changes: 

```bash
mysql> show variables like "table_open_cache";
+------------------+-------+
| Variable_name    | Value |
+------------------+-------+
| table_open_cache | 2000  |
+------------------+-------+
1 row in set (0.01 sec)

mysql> show variables like "wait_timeout";
+---------------+-------+
| Variable_name | Value |
+---------------+-------+
| wait_timeout  | 28800 |
+---------------+-------+
1 row in set (0.01 sec)

mysql> show variables like "connect_timeout";
+-----------------+-------+
| Variable_name   | Value |
+-----------------+-------+
| connect_timeout | 10    |
+-----------------+-------+
1 row in set (0.00 sec)
```

For others that might be interested, the timeout values are in seconds, not milliseconds or anything. Here are the docs: 

[`table_open_cache`](https://dev.mysql.com/doc/refman/5.6/en/server-system-variables.html#sysvar_table_open_cache)
[`wait_timeout`](https://dev.mysql.com/doc/refman/5.6/en/server-system-variables.html#sysvar_wait_timeout)
[`connect_timeout`](https://dev.mysql.com/doc/refman/5.6/en/server-system-variables.html#sysvar_connect_timeout)

The two that stood out to me as culprits are `connect_timeout` being 10 seconds and `table_open_cache` being 2000. `table_open_cache` is: 

> The number of open tables for all threads. 

With our Mac Pro (aka "The Beast") there are 2 Xeon processors, each with 6 cores and 12 threads, thus resulting in 24 threads. 

If each thread is opening up 100 tables to run the test suite, that's 2,400 open tables across all the threads, excluding any system tables that might be open or other activity with the MySQL daemon.

So, with all configuration limits, we could do some math or just double it until things work. 

Here's what I put into my `/etc/my.cnf` file: 

```text
[mysqld]
max_connections=100000
max_connect_errors=100000
connect_timeout=60
table_open_cache=4000
```

Just to confirm after rebooting MySQL: 

```bash
mysql> show variables like "table_open_cache";
+------------------+-------+
| Variable_name    | Value |
+------------------+-------+
| table_open_cache | 4000  |
+------------------+-------+
1 row in set (0.01 sec)

mysql> show variables like "connect_timeout";
+-----------------+-------+
| Variable_name   | Value |
+-----------------+-------+
| connect_timeout | 60    |
+-----------------+-------+
1 row in set (0.00 sec)
```

Btw, you can monitor these variables whilst running your test suite by using `show status;`: 

```bash
mysql> show status like "Open%";
+--------------------------+-------+
| Variable_name            | Value |
+--------------------------+-------+
| Open_files               | 19    |
| Open_streams             | 0     |
| Open_table_definitions   | 570   |
| Open_tables              | 591   |
| Opened_files             | 11594 |
| Opened_table_definitions | 0     |
| Opened_tables            | 1     |
+--------------------------+-------+
7 rows in set (0.00 sec)
```

- - -

**3 Hours Later...**

These settings weren't helping so at this point I decided to check the error logs and found lots of these two: 

```
[Note] Aborted connection 271 to db: 'unconnected' user: 'root' host: 'localhost' (Got an error reading communication packets)
```

```
[Warning] File Descriptor 1234 exceeded FD_SETSIZE=1024
```

The first one ended up being a total red herring. I wasted hours trying to track down what was going on here, assuming this was the culprit, but it actually showed that message even when I ran my test suite with just a single thread and everything went fine. Plus, it only showed up when the thread was finished, i.e. disconnecting from the database, instead of at the beginning or in the middle. Very odd.

The second one is the real issue. I won't go into great detail here because it's now late and I went down the rabbit hole pretty hard, but `FD_SETSIZE` is essentially a hard coded file descriptor limit in the OS. You can't just increase it like you would with an open file limit or something.

The way to solve it is actually by _reducing_ your `table_open_cache` instead of increasing it. The `table_open_cache` keeps a file descriptor for each table cache it has so if this value is big and you have lots of concurrent connections, you start to exceed the `FD_SETSIZE` limit.

I found the answer from "Phil" over [here](https://www.simplemachines.org/community/index.php?topic=558754.0):

> The default setting for table-open-cache was 2000, which was greater than the MacOS limit.  Decreasing this to 400 has fixed my database error problems without any noticeable averse effects.

The gotcha is that the default value is _too high_ instead of _too low_, which is not intuitive. I never would have figured that out. Thank you, Phil!

I made a number of other changes to the `/etc/my.cnf` configuration while tracking this down so I'm going to remove them one by one and re-run my test suite to see if it works with just this single change...

- - - 

I was able to remove all of my other values in `/etc/my.cnf` without seeing anymore errors.

My final `/etc/my.cnf` file looks like this now: 

```text
[mysqld]
table_open_cache=900
```

You can play with the `table_open_cache` number a bit to see where it hits the limit based on your machine and needs but that'll be an exercise I leave to the reader. We found that anything equal to or greater than `1000` would cause errors, likely because of the `FD_SETSIZE` limit is 1024.

**Comment by baljeet03** [comment]
Thats why I asked for `max_connections` :). I thought you might have `max_connections` somewhere around 100 or 200(mine is 150). Though your `max_connections` is way too high. How much RAM is it consuming? At least you are not using mysql 5.6 which consume memory for all the connection during setup time instead of usage. 
I would suggest to use `max_connections=1000-1500` depending on your load and `table_open_cache` to be 2x to 5x of `max_connections`. You can always increase the `file_descriptor` size with little bit googling. Use threadpool provided by mysql.

**Comment by joshuapinter** [comment]
Thanks @baljeet03. After getting rid of all the test code, my `max_connections` is set to 1000. 

I couldn't find an easy way to upping the `FD_SETSIZE` without getting deep into things so I'll just keep the `table_open_cache` at a slightly lesser amount so I don't bump into that ceiling.

Thanks again! 👍

**Comment by baljeet03** [comment]
My pleasure @joshuapinter. Regarding 'table_open_cache', check the load basically considering tables sizes and set the values accordingly. I personally don't think you will face any problem if you set it 2x or 3x if you have RAM more than 8 GB and having huge tables.

**Comment by joshuapinter** [comment]
I found that I have to reduce the `table_open_cache` because as soon as my `Open_tables` value from `show status like "Open%";` hits 1000, I exceed the `FD_SETSIZE` limit.

My `max_connections` is set to 1000 but I'm not sure how many are actually being used. I'll see if I can test that and get the value from `show status;`...

**Comment by joshuapinter** [comment]
Okay, just tested it and the value for `Connections` maxed out at around 157 during the test suite and the `Open_tables` and `Open_table_definitions` hovered around 900 each.

If I bump up the `table_open_cache` to 1000, it fails to connect because it hits that `FD_SETSIZE` limit.

Hopefully that helps shed light on my environment and what's happening.

**Comment by sunwei415** [comment]
set table_open_cache to 400 and restarted MySQL. Error gone.

MySQL 5.6 @ macOS Big Sur 11.1

**Comment by alexventuraio** [comment]
> set table_open_cache to 400 and restarted MySQL. Error gone.
> 
> MySQL 5.6 @ macOS Big Sur 11.1.
> 
> Maybe a sole restarting would do the trick as well.

In my case, I tried to change `table_open_cache` from `2000` (default value) to `400` and it did not work.
Then, just restarting my computer fixed the problem. 🙄

**Comment by RichStone** [comment]
> > set table_open_cache to 400 and restarted MySQL. Error gone.
> > MySQL 5.6 @ macOS Big Sur 11.1.
> > Maybe a sole restarting would do the trick as well.
> 
> In my case, I tried to change `table_open_cache` from `2000` (default value) to `400` and it did not work. Then, just restarting my computer fixed the problem. 🙄

Just restart fixed it for me

**Comment by salochara** [comment]
> > > set table_open_cache to 400 and restarted MySQL. Error gone.
> > > MySQL 5.6 @ macOS Big Sur 11.1.
> > > Maybe a sole restarting would do the trick as well.
> > 
> > 
> > In my case, I tried to change `table_open_cache` from `2000` (default value) to `400` and it did not work. Then, just restarting my computer fixed the problem. 🙄
> 
> Just restart fixed it for me

Same for me. Restarted ny computer and "magically" started working 🙄
